Unified Unit Category Checking For Uniques

givehub99 opened the pull request:
Starting unified unit category checking. This will increase flexibility with uniques as well as cut down on uniques that have only a slightly different way of checking unit categories.

Things currently checked by the category checking method are:
Type of unit 
Name of unit
Whether unit is wounded
If the unit is land, water, or air

"Bonus vs City y%" still works currently.

Uniques that have been changed/added to use this category checking method so far are: 

"Bonus vs x y%"
"[] units deal +[]% damage" (parameterization of "Wounded military units deal +25% damage")
"[] units gain the [] promotion" 

If PR approved, using this method to change/upgrade more uniques would be a good idea.

Commented as deprecated are these uniques
"+[]% Strength vs []" (this has not been used in json files)
"Wounded military units deal +25% damage" (policies json file was updated to use parameterized version)

The following was added to template.properties for translating modifiers:

Air
Wounded

yairm210 commented on core/src/com/unciv/logic/battle/CityCombatant.kt:
```diff
@@ -62,5 +62,13 @@ class CityCombatant(val city: CityInfo) : ICombatant {
         return strength.roundToInt()
     }
 
+    override fun matchesCategory(category:String): Boolean {
```
override fun matchesCategory(category:String) = category == "City"

yairm210 commented on core/src/com/unciv/logic/battle/MapUnitCombatant.kt:
```diff
@@ -38,4 +38,12 @@ class MapUnitCombatant(val unit: MapUnit) : ICombatant {
         return unit.name+" of "+unit.civInfo.civName
     }
+
+    override fun matchesCategory(category:String): Boolean {
```
override fun matchesCategory(category:String) = unit.matchesCategory(category)

yairm210 commented on core/src/com/unciv/logic/map/MapUnit.kt:
```diff
@@ -696,5 +696,24 @@ class MapUnit {
         }
     }
 
+    fun matchesCategory(category:String): Boolean {
```
DO NOT allow lowercasing - the translation is case-sensitive, and so the uniques should be too

yairm210 commented on core/src/com/unciv/logic/map/MapUnit.kt:
```diff
@@ -696,5 +696,24 @@ class MapUnit {
+    fun matchesCategory(category:String): Boolean {
+        val truncatedCategory = category.replace(" units", "").toLowerCase()
+
+        if (truncatedCategory == type.toString().toLowerCase())
```
type.name

yairm210 commented:
Overall looks good! 
This is one of the three unifications that's been needed for a while: Tile checking (done), unit checking (this) and city checking (less important tbh)

givehub99 commented:
> 
> 
> Overall looks good!
> This is one of the three unifications that's been needed for a while: Tile checking (done), unit checking (this) and city checking (less important tbh)

-Changed matchesCategory method for CityCombatant to what was suggested
-Changed matchesCategory method for MapUnitCombatant to what was suggested
-Use type.name instead of type.toString().toLowerCase()
-Case mixing is no longer allowed. For wounded, land, water or air, you have to either use the capitalized version (ie. "Wounded") or lower case with units (ie. "wounded units"). This lets preexisting uniques that use "wounded units" etc still work while also letting people use the singular capitalized word (ie. "Wounded") for other uniques. All of these versions are already accounted for in template.properties for translations

yairm210 commented on core/src/com/unciv/logic/battle/BattleDamage.kt:
```diff
@@ -45,18 +45,13 @@ object BattleDamage {
+
             }
 
+            // As of 3.11.1 This is to be deprecated and converted to "Bonus vs x y%" - keeping it here so that mods with this can still work for now
```
Au Contrair, I want to convert all the Bonus vs x y% TO +[]% Strength vs [] ;)

Better because of:
- Unified unique handling
- Less regex use
- No need for custom translation

yairm210 replied:
Not that I think a *comment* is a good enough reason to avoid merging this, just FYI ;)
